eclair-core/src/main/scala/fr/acinq/eclair/channel/TxPublisher.scala

```diff
@@ -0,0 +1,418 @@
+      }
+    } else {
+      unsignedTx match {
+        case htlcSuccess: HtlcSuccessTx => htlcSuccess.copy(tx = htlcSuccess.tx.copy(txOut = Seq(htlcSuccess.tx.txOut.head)))
+        case htlcTimeout: HtlcTimeoutTx => htlcTimeout.copy(tx = htlcTimeout.tx.copy(txOut = Seq(htlcTimeout.tx.txOut.head)))
+      }
+    }
+  }
+
+  sealed trait HtlcTxAndWitnessData {
+    // @formatter:off
+    def txInfo: HtlcTx
+    def updateTx(tx: Transaction): HtlcTxAndWitnessData
+    def addSigs(localSig: ByteVector64, commitmentFormat: CommitmentFormat): HtlcTx
+    // @formatter:on
+  }
+
+  object HtlcTxAndWitnessData {
+
+    case class HtlcSuccess(txInfo: HtlcSuccessTx, remoteSig: ByteVector64, preimage: ByteVector32) extends HtlcTxAndWitnessData {
+      // @formatter:off
+      override def updateTx(tx: Transaction): HtlcTxAndWitnessData = copy(txInfo = txInfo.copy(tx = tx))
+      override def addSigs(localSig: ByteVector64, commitmentFormat: CommitmentFormat): HtlcTx = Transactions.addSigs(txInfo, localSig, remoteSig, preimage, commitmentFormat)
+      // @formatter:on
+    }
+
+    case class HtlcTimeout(txInfo: HtlcTimeoutTx, remoteSig: ByteVector64) extends HtlcTxAndWitnessData {
+      // @formatter:off
+      override def updateTx(tx: Transaction): HtlcTxAndWitnessData = copy(txInfo = txInfo.copy(tx = tx))
+      override def addSigs(localSig: ByteVector64, commitmentFormat: CommitmentFormat): HtlcTx = Transactions.addSigs(txInfo, localSig, remoteSig, commitmentFormat)
+      // @formatter:on
+    }
+
+    def apply(txInfo: HtlcTx, commitments: Commitments): Option[HtlcTxAndWitnessData] = {
+      txInfo match {
+        case tx: HtlcSuccessTx =>
+          commitments.localChanges.all.collectFirst {
+            case u: UpdateFulfillHtlc if Crypto.sha256(u.paymentPreimage) == tx.paymentHash => u.paymentPreimage
+          }.flatMap(preimage => {
+            commitments.localCommit.publishableTxs.htlcTxsAndSigs.collectFirst {
+              case HtlcTxAndSigs(HtlcSuccessTx(input, _, _, _), _, remoteSig) if input.outPoint == tx.input.outPoint => HtlcSuccess(tx, remoteSig, preimage)
+            }
+          })
+        case tx: HtlcTimeoutTx =>
+          commitments.localCommit.publishableTxs.htlcTxsAndSigs.collectFirst {
+            case HtlcTxAndSigs(HtlcTimeoutTx(input, _, _), _, remoteSig) if input.outPoint == tx.input.outPoint => HtlcTimeout(tx, remoteSig)
+          }
+      }
+    }
+
+  }
+
+}
+
+private class TxPublisher(nodeParams: NodeParams, watcher: akka.actor.ActorRef, client: ExtendedBitcoinClient, context: ActorContext[TxPublisher.Command]) {
+
+  import TxPublisher._
+
+  private case class TxWithRelativeDelay(childTx: PublishTx, parentTxIds: Set[ByteVector32])
+
+  val log = context.log
+  val keyManager = nodeParams.channelKeyManager
+  val feeEstimator = nodeParams.onChainFeeConf.feeEstimator
+  val feeTargets = nodeParams.onChainFeeConf.feeTargets
```

Comment:
```suggestion
  private val log = context.log
  private val keyManager = nodeParams.channelKeyManager
  private val feeEstimator = nodeParams.onChainFeeConf.feeEstimator
  private val feeTargets = nodeParams.onChainFeeConf.feeTargets
```

It feels weird to redefine those as `val`, I think `def` would be a better fit.

I would even go with:
```suggestion
  import context.log
  import nodeParams.{channelKeyManager => keyManager}
  import nodeParams.onChainFeeConf.feeEstimator
  import nodeParams.onChainFeeConf.feeTargets
```

Reply:
Done in https://github.com/ACINQ/eclair/pull/1749/commits/d17b3d0ab1d33f504c04ad67a537b46649ee3230 except for `context.log` which somehow freezes the tests...?